# Linear Programmin - M2
- https://www.scaler.com/topics/linear-programming-in-python/

In [1]:
# importing libraries and modules   
import numpy as np
from scipy.optimize import linprog
import matplotlib.pyplot as plt
from matplotlib.patches import Polygon

In [ ]:
print("This problem statmeent can be formulated as:")

print("max 3a + 4b")
print("this must be subjected to:")
print("2a + 5b <= 30")
print("4a + 2b <= 20")
print("a,b >0")

In [2]:
# Constructing the parameters
x = np.array([3, 4])
# Below are the Inequality constraints
D = np.array([[2, 5], [4, 2]])
E = np.array([30,20])

In [4]:
# Solving the linear programming in Python problem
# As linprog does minimization, Purposely did put a negative sign on the objective 
result = linprog(-x, A_ub=D, b_ub=E, method='highs')
# As the problem wants to maximize the objective, we need to put a minus sign in front of parameter vector c.
print(result)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -27.5
              x: [ 2.500e+00  5.000e+00]
            nit: 2
          lower:  residual: [ 2.500e+00  5.000e+00]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00  0.000e+00]
                 marginals: [-6.250e-01 -4.375e-01]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0


- As seen above, the optimal plan comes out as 2.5, 5 which tells the organization for producing 2.5 units of Product S and 5 units of Product T. The maximized revenue that could be generated is 27.5. Here, the linprog function is a black box where Python first transforms the problem as a standard form. It evaluates each inequality constraint it generates using one slack variable.